platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/LongListDisk.java

```diff
@@ -16,22 +16,76 @@
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.CopyOnWriteArrayList;
 
-/** A direct on disk implementation of LongList */
-public class LongListDisk extends LongList {
+/**
+ *  A direct on disk implementation of LongList. This implementation creates a temporary file to store the data.
+ *  If the user provides a file, it will be used to create a copy of the in the temporary file. If the file is absent,
+ *  it will take the name of the file provided by the user and create a temporary file with this name.
+ * <p>
+ *  Unlike the "snapshot" file, the temporary files doesn't contain the header, only the body.
+ *
+ */
+public class LongListDisk extends LongList<Long> {
+
+    private static final String STORE_POSTFIX = "longListDisk";
+    private static final String DEFAULT_FILE_NAME = "LongListDisk.ll";
     /** A temp byte buffer for reading and writing longs */
-    private static final ThreadLocal<ByteBuffer> TEMP_LONG_BUFFER_THREAD_LOCAL =
-            ThreadLocal.withInitial(() -> ByteBuffer.allocateDirect(Long.BYTES).order(ByteOrder.nativeOrder()));
-    /** The disk file this LongList is based on */
-    private final Path file;
+    private final ThreadLocal<ByteBuffer> tempLongBufferThreadLocal;
+
+    /** This file channel is to work with the temporary file. The field is effectively immutable, however it can't be
+     * declared final because in some cases it has to be initialized in {@link LongListDisk#readBodyFromFileChannelOnInit}
+     */
+    private FileChannel currentFileChannel;
+
+    /** A temp byte buffer for transferring data between file channels */
+    private final ThreadLocal<ByteBuffer> transferBufferThreadLocal;
+
+    /**
+     * Offsets of the chunks that are free to be used. The offsets are relative to the start of the file.
+     */
+    private final Deque<Long> freeChunks;
+
+    /**
+     * A list of all the byte buffers that are used to read and write data from the file. This is used to ensure that
+     * the buffers are properly disposed of when the list is closed.
+     */
+    private final List<ByteBuffer> bufferRegistry;
+
+    {
+        transferBufferThreadLocal = createThreadLocalByteBuffer(memoryChunkSize);
+        tempLongBufferThreadLocal = createThreadLocalByteBuffer(Long.BYTES);
+        bufferRegistry = new CopyOnWriteArrayList<>();
+        freeChunks = new ConcurrentLinkedDeque<>();
+        try {
+            if (currentFileChannel == null) {
+                currentFileChannel = FileChannel.open(
+                        createTempFile(DEFAULT_FILE_NAME),
+                        StandardOpenOption.CREATE,
+                        StandardOpenOption.READ,
+                        StandardOpenOption.WRITE);
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public LongListDisk() {
+        super(DEFAULT_NUM_LONGS_PER_CHUNK, DEFAULT_MAX_LONGS_TO_STORE, DEFAULT_RESERVED_BUFFER_LENGTH);
```

Comment:
Should it be 1 instead of DEFAULT_NUM_LONGS_PER_CHUNK? LongListDisk chunks are individual longs according to the class type.

Reply:
The notion of `chunks` is consistent across the implementations. A `chunk` is a unit consisting of multiple long values. It can be represented as an array (`LongListHeap`), as a `ByteBuffer` (`LongListOffHeap`) or as a long value which is an offset of reserved space in a file (`LongListDisk`). 

So, the parameter `DEFAULT_NUM_LONGS_PER_CHUNK` means exactly the same thing for all the implementation: the number of longs that can fit into a chunk.